### PR TITLE
Refactor Compare to be more modular and extensible

### PR DESCRIPTION
**Pull request by robieta**
Right now `Compare` is kind of a mess of comprehensions and string manipulation, and extending it is quite hard. (For instance, even adding the ability to print callgrind results would be a non-trivial amount of work.) This PR replaces the old implementation with a much more composable implementation:

1) Cells can produce arbitrary output (including newlines, ANSI color codes, etc.) based on what is appropriate. (And have reduction methods to coordinate with adjacent values.)

2) The Table handles layout and alignment so that cell implementers don't have to concern themselves with it.

Not only does it make the Compare source much more readable (and also now MyPy strict compliant), but it means that users have a lot more control over how to lay out the data. (For instance we go over a bandwidth example in the new README.) I moved the source to a `visualize` internal namespace just to try to keep things tidy.

Test plan: Existing unit tests should suffice. The layout changes slightly, but in my opinion for the better.

I've added Peter, Alexander, and Bert as "power user reviewers"; I'm curious to hear if this would make your life easier.

### CODE REVIEW

**Comment by robieta** [comment]
P.S. This is basically subsuming the display part of https://github.com/pytorch/pytorch/pull/50610

**Comment by robieta** [comment]
I'm not sure why MyPy is having issues. Maybe because it's a new path?

**Comment by robieta** [comment]
Gah! Missing a comma that was causing two paths to be concatenated in the MyPy ini file.

File: torch/utils/benchmark/visualize/table.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+    @property
+    def should_render(self) -> Layout.ShouldRender:
+        return Layout.ShouldRender.YES
+
+    def set_context(self, row: "Row") -> None:
+        # We don't want to rely on __init__, as row is not known at construction.
+        self._row = row
+
+    def row_reduce(self, f: REDUCTION_FUNCTION) -> REDUCTION_OUTPUT:
```
Comment (ngimel) [inline torch/utils/benchmark/visualize/table.py:92]: as these methods are expected to be used by end users in subclassed DataCells, they deserve to be documented

File: torch/utils/benchmark/visualize/table.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        return self._row._table._all_reduce(self, f)
+
+
+class FullRowCell(Cell):
+    """This type of cell expects to occupy an entire row of a Table.
+
+    Consequently, it also takes a `width` argument to `render` to that it can
+    decide how to fill the row.
+
+    All `FullRowCell` instances are considered to occupy a separate column.
```
Comment (ngimel) [inline torch/utils/benchmark/visualize/table.py:111]: are considered to occupy the same column?

File: torch/utils/benchmark/utils/common.py [hunk 3]
```diff
@@ -276,6 +277,12 @@ def trim_sigfig(x: float, n: int) -> float:
     return float(np.round(x / scale) * scale)
 
 
+def get_unique(values: Iterable[T]) -> T:
```
Comment (ngimel) [inline torch/utils/benchmark/utils/common.py:280]: it's more like `assert_all_equal`?

**Review by ngimel — COMMENTED** [review]
This is a cool refactor, and the ability to compute and display dependent metrics is great!
I've left a few comments, mostly about documentation.

**Pull request closed without merging**